Modules/_decimal/_decimal.c

```diff
@@ -3672,34 +3676,45 @@ dec_as_long(PyObject *dec, PyObject *context, int round)
     }
 
     status = 0;
-    ob_digit = NULL;
+
+    int64_t val = mpd_qget_i64(x, &status);
+    if (!status) {
+        mpd_del(x);
+        return PyLong_FromInt64(val);
+    }
+
+    const uint8_t bpd = layout->bits_per_digit;
+    n = (mpd_sizeinbase(x, 2) + bpd - 1) / bpd;
+    PyLongWriter *writer = PyLongWriter_Create(mpd_isnegative(x), n,
+                                               (void**)&ob_digit);
+    /* mpd_sizeinbase can overestimate size by 1 digit, set it to zero. */
+    ob_digit[n-1] = 0;
+    if (writer == NULL) {
+        PyLongWriter_Discard(writer);
```

Comment:
You should not call PyLongWriter_Discard() if PyLongWriter_Create() failed.

Currently, calling PyLongWriter_Discard(NULL) does crash. Maybe we should modify PyLongWriter_Discard() to accept NULL. @serhiy-storchaka suggested that if I recall correctly.